File: Core/Architect.API.Core.DataAccess/Security/UserMember.vb
Imports Architect.Common.Proxy

Namespace Architect.API.Core.DataAccess.Security

    Partial Public NotInheritable Class UserMember


        ''' <summary>
        ''' Actualiza un registro en la tabla UserMember por medio de su clave primaria.
        ''' </summary>
        ''' <param name="usermemberItem">Instancia de UserMember</param>
        ''' <returns>Cantidad de registros actualizados.</returns>
        Public Shared Function InternalUpdate(usermemberItem As Architect.API.Core.Contracts.Security.UserMember) As Integer
            Dim rows As Integer = 0
            With New DataManagerFactory("UPDATE UserMember " &
                                           "SET Password=@:Password, OldPassword=@:OldPassword, FailedPasswordCount=@:FailedPasswordCount, SecurityLevel=@:SecurityLevel, IsLockedOut=@:IsLockedOut, LockedOutDate=@:LockedOutDate, PasswordChangedDate=@:PasswordChangedDate, OneTimePassword=@:OneTimePassword, LoginDate=@:LoginDate " &
                                         "WHERE UserId=@:UserId", "UserMember", "Research")
                .AddParameter("Password", DbType.AnsiString, 255, usermemberItem.Password.IsEmpty(), usermemberItem.Password)
                .AddParameter("OldPassword", DbType.AnsiString, 255, usermemberItem.OldPassword.IsEmpty(), usermemberItem.OldPassword)
                .AddParameter("FailedPasswordCount", DbType.Decimal, 3, usermemberItem.FailedPasswordCount.IsEmpty(), usermemberItem.FailedPasswordCount)
                .AddParameter("SecurityLevel", DbType.Decimal, 3, False, usermemberItem.SecurityLevel)
                .AddParameter("IsLockedOut", DbType.Decimal, 1, False, IIf(usermemberItem.IsLockedOut, 1, 0))
                .AddParameter("LockedOutDate", DbType.DateTime, 9, usermemberItem.LockedOutDate.IsEmpty(), usermemberItem.LockedOutDate)
                .AddParameter("PasswordChangedDate", DbType.DateTime, 9, usermemberItem.PasswordChangedDate.IsEmpty(), usermemberItem.PasswordChangedDate)
                .AddParameter("OneTimePassword", DbType.AnsiStringFixedLength, 6, usermemberItem.OneTimePassword.IsEmpty(), usermemberItem.OneTimePassword)
                .AddParameter("LoginDate", DbType.DateTime, 9, usermemberItem.LoginDate.IsEmpty(), usermemberItem.LoginDate)
                .AddParameter("UserId", DbType.Decimal, 9, False, usermemberItem.UserId)
                rows = .CommandExecute()
            End With
            Return rows
        End Function

        ''' <summary>
        ''' Recupera una lista con la información de correo electronico de los usuario asociados a un rol.
        ''' </summary>
        ''' <param name="companyId">Identificación de la compañia propietaria.</param>
        ''' <param name="roleName">Nombre del rol usando para filtrar la lista</param>
        ''' <returns>Lista de instancias de UserMember</returns>
        Public Shared Function EmailInfoByRoleName(companyId As Integer, roleName As String) As List(Of Architect.API.Core.Contracts.Security.UserMember)
            Dim rows As Data.DataTable = Nothing
            Dim result As New List(Of Architect.API.Core.Contracts.Security.UserMember)

            roleName &= ",,,"

            With New DataManagerFactory(My.Resources.UserMember_EmailInfoByRoleName, "UserMember", "Research")
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyId)
                .AddParameter("RoleName1", DbType.AnsiString, 80, False, roleName.Split(",")(0).Trim().ToLower())
                .AddParameter("RoleName2", DbType.AnsiString, 80, False, roleName.Split(",")(1).Trim().ToLower())
                .AddParameter("RoleName3", DbType.AnsiString, 80, False, roleName.Split(",")(2).Trim().ToLower())

                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    For Each itemRow As DataRow In rows.Rows
                        result.Add(New Contracts.Security.UserMember With {.EMail = itemRow.StringValue("EMail"),
                                                                           .FirstName = itemRow.StringValue("FirstName"),
                                                                           .LastName = itemRow.StringValue("LastName")})
                    Next
                End If
            End With
            Return result
        End Function

        ''' <summary>
        ''' Recupera un registro en la tabla UserMember por medio de su clave primaria.
        ''' </summary>
        ''' <param name="eMail">E-Mail del Usuario.</param>
        ''' <param name="companyId">Identificación de la compañia propietaria.</param>
        ''' <returns>Instancia de UserMember</returns>
        Public Shared Function RetrieveByEMail(eMail As String, companyId As Integer) As Architect.API.Core.Contracts.Security.UserMember
            Dim rows As Data.DataTable = Nothing
            Dim result As Architect.API.Core.Contracts.Security.UserMember = Nothing
            With New DataManagerFactory("SELECT UserId, UserName, EMail, Password, OldPassword, FirstName, LastName, FailedPasswordCount, SecurityLevel, IsLockedOut, LockedOutDate, OneTimePassword, PasswordChangedDate, ManagerId, RecordStatus, CompanyId, LoginDate, BranchOffice, IdentificationType, Identification " &
                                          "FROM UserMember " &
                                         "WHERE LOWER(EMail)=@:EMail AND CompanyId=@:CompanyId", "UserMember", "Research")
                .AddParameter("EMail", DbType.AnsiString, 120, False, eMail.IfEmpty(String.Empty).ToLower.Trim)
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyId)
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    result = DataRowMapperLight(rows.FirstRow)
                End If
            End With
            Return result
        End Function

        ''' <summary>
        ''' Recupera un registro en la tabla UserMember por medio de su clave primaria.
        ''' </summary>
        ''' <param name="userName">Nombre Usuario.</param>
        ''' <param name="companyId">Identificación de la compañia propietaria.</param>
        ''' <returns>Instancia de UserMember</returns>
        Public Shared Function RetrieveByUserName(userName As String, companyId As Integer) As Architect.API.Core.Contracts.Security.UserMember
            Dim rows As Data.DataTable = Nothing
            Dim result As Architect.API.Core.Contracts.Security.UserMember = Nothing
            With New DataManagerFactory("SELECT UserId, UserName, EMail, Password, OldPassword, FirstName, LastName, FailedPasswordCount, SecurityLevel, IsLockedOut, LockedOutDate, OneTimePassword, PasswordChangedDate, ManagerId, RecordStatus, CompanyId, LoginDate, BranchOffice, IdentificationType, Identification " &
                                          "FROM UserMember " &
                                         "WHERE LOWER(UserName)=@:UserName AND CompanyId=@:CompanyId", "UserMember", "Research")
                .AddParameter("UserName", DbType.AnsiString, 35, False, userName.IfEmpty(String.Empty).ToLower.Trim)
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyId)
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    result = DataRowMapperLight(rows.FirstRow)
                End If
            End With
            Return result
        End Function


        ''' <summary>
        ''' Recupera un registro en la tabla UserMember por medio de una clave de acceso.
        ''' </summary>
        ''' <param name="accessKey">valor usado para identificar un usuario para servicios.</param>
        ''' <returns>Instancia de UserMember</returns>
        Public Shared Function RetrieveByAccessKey(accessKey As String) As Architect.API.Core.Contracts.Security.UserMember
            Dim rows As Data.DataTable = Nothing
            Dim result As Architect.API.Core.Contracts.Security.UserMember = Nothing
            With New DataManagerFactory("SELECT UserId, UserName, EMail, Password, OldPassword, FirstName, LastName, FailedPasswordCount, SecurityLevel, IsLockedOut, LockedOutDate, OneTimePassword, PasswordChangedDate, ManagerId, RecordStatus, CompanyId, LoginDate, BranchOffice, IdentificationType, Identification " &
                                          "FROM UserMember " &
                                         "WHERE AccessKey=@:AccessKey", "UserMember", "Research")
                .AddParameter("AccessKey", DbType.AnsiString, 20, False, accessKey)
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    result = DataRowMapperLight(rows.FirstRow)
                End If
            End With
            Return result
        End Function

        ''' <summary>
        ''' Permite hacer el mapeo entre una fila de un datatable hacia una instancia de UserMember.
        ''' </summary>
        ''' <returns>Instancia de UserMember</returns>
        Private Shared Function DataRowMapperLight(itemRow As DataRow) As Architect.API.Core.Contracts.Security.UserMember
            Return New Architect.API.Core.Contracts.Security.UserMember With {.UserId = itemRow.NumericValue("UserId"), .UserName = itemRow.StringValue("UserName"), .EMail = itemRow.StringValue("EMail"), .Password = itemRow.StringValue("Password"), .OldPassword = itemRow.StringValue("OldPassword"), .FirstName = itemRow.StringValue("FirstName"), .LastName = itemRow.StringValue("LastName"), .FailedPasswordCount = itemRow.NumericValue("FailedPasswordCount"), .SecurityLevel = itemRow.NumericValue("SecurityLevel"), .IsLockedOut = (itemRow.NumericValue("IsLockedOut") = 1), .PasswordChangedDate = itemRow.DateTimeValue("PasswordChangedDate"), .ManagerId = itemRow.NumericValue("ManagerId"), .RecordStatus = itemRow.NumericValue("RecordStatus"), .CompanyId = itemRow.NumericValue("CompanyId"), .LockedOutDate = itemRow.DateTimeValue("LockedOutDate"), .OneTimePassword = itemRow.StringValue("OneTimePassword"), .LoginDate = itemRow.DateTimeValue("LoginDate"), .BranchOffice = itemRow.NumericValue("BranchOffice"), .IdentificationType = itemRow.NumericValue("IdentificationType"), .Identification = itemRow.StringValue("Identification")}
        End Function

    End Class

End Namespace

File: Core/Architect.API.Core.DataAccess/Security/UserMember.cs
using Architect.DataFactory;
using System;
using System.Data;

namespace Architect.API.Core.DataAccess.Security
{
    public sealed partial class UserMember
    {
        ///// <summary>
        ///// Recupera una lista de registros en la tabla UserMember.
        ///// </summary>
        ///// <param name="companyId">Identificación de la compañía propietaria.</param>
        ///// <param name="filter">Filtro personalizado.</param>
        ///// <param name="recordStatus"></param>
        ///// <returns>Lista de instancias de UserMember</returns>
        //public static List<Contracts.Security.UserMember> RetrieveAll(int companyId, int securityLevel, string filter, string recordStatus, IDbConnection connection = null)
        //{
        //    List<Architect.API.Core.Contracts.Security.UserMember> result = new List<Architect.API.Core.Contracts.Security.UserMember>();
        //    Database.Select(@"SELECT UserMember.UserId, UserMember.CompanyId, UserMember.UserName, UserMember.EMail, UserMember.Password, UserMember.OldPassword, UserMember.IdentificationType, UserMember.Identification, UserMember.FirstName, UserMember.LastName, UserMember.BirthDate, UserMember.FailedPasswordCount, UserMember.SecurityLevel, UserMember.IsLockedOut, UserMember.LockedOutDate, UserMember.PasswordChangedDate, UserMember.OneTimePassword, UserMember.LoginDate, UserMember.ManagerId, UserMember.AccessKey, UserMember.BranchOffice, UserMember.Reference, UserMember.Position, UserMember.PhoneNumber, UserMember.SalesChannel, UserMember.CustomData, UserMember.RecordStatus, UserMember.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, UserMember.UpdateDate,
        //                            (SELECT LISTAGG(RM.ROLENAME , ', ') WITHIN GROUP (ORDER BY RM.ROLENAME )
        //                       FROM USERROLEMEMBER urm LEFT JOIN ROLEMEMBER rm ON rm.RoleId = urm.RoleId WHERE urm.UserId = UserMember.UserId) Roles
        //                       FROM UserMember LEFT JOIN UserMember um ON um.UserId = UserMember.UpdateUserCode
        //                      WHERE UserMember.CompanyId=:CompanyId AND UserMember.SecurityLevel<=:SecurityLevel" + BuildFilter(filter, recordStatus))
        //                .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
        //                .AddParameter("SecurityLevel", DbType.Decimal, 3, securityLevel)
        //                .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
        //                {
        //                    Architect.API.Core.Contracts.Security.UserMember item = DataReaderToUserMember(reader);
        //                    item.RoleList = reader.StringValue("Roles");
        //                    result.Add(item);
        //                }));
        //    return result;
        //}

        //public static string BuildFilter(string filter, string recordStatus)
        //{
        //    string result = string.Empty;
        //    if (filter.IsNotEmpty())
        //    {
        //        result = " AND (UPPER(UserMember.UserName) LIKE '%" + filter.ToUpper() + "%' " + "OR UPPER(UserMember.FirstName) LIKE '%" + filter.ToUpper() + "%' " + "OR UPPER(UserMember.LastName) LIKE '%" + filter.ToUpper() + "%' " + "OR UPPER(UserMember.EMail) LIKE '%" + filter.ToUpper() + "%') ";
        //    }

        //    if (recordStatus.IsNotEmpty())
        //    {
        //        result = " AND UserMember.RecordStatus IN (" + recordStatus + ") ";
        //    }

        //    return result;
        //}

        ///// <summary>
        ///// Recupera una lista con la información de correo electrónico de los usuario asociados a un rol.
        ///// </summary>
        ///// <param name="companyId">Identificación de la compañía propietaria.</param>
        ///// <param name="roleName">Nombre del rol usando para filtrar la lista</param>
        ///// <returns>Lista de instancias de UserMember</returns>
        //public static List<Contracts.Security.UserMember> EmailInfoByRoleName(int companyId, string roleName, IDbConnection connection = null)
        //{
        //    List<Architect.API.Core.Contracts.Security.UserMember> result = new List<Architect.API.Core.Contracts.Security.UserMember>();
        //    roleName += ",,,";

        //    Database.Select(Properties.Resources.UserMember_EmailInfoByRoleName)
        //        .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
        //        .AddParameter("RoleName1", DbType.AnsiString, 80, roleName.Split(',')[0].Trim().ToLower())
        //        .AddParameter("RoleName2", DbType.AnsiString, 80, roleName.Split(',')[1].Trim().ToLower())
        //        .AddParameter("RoleName3", DbType.AnsiString, 80, roleName.Split(',')[2].Trim().ToLower())
        //        .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
        //        {
        //            result.Add(new Contracts.Security.UserMember()
        //            {
        //                EMail = reader.StringValue("EMail"),
        //                FirstName = reader.StringValue("FirstName"),
        //                LastName = reader.StringValue("LastName")
        //            });
        //        }));
        //    return result;
        //}

        /// <summary>
        /// Último valor asignado a clave única de la tabla UserMember.
        /// </summary>
        /// <param name="connection">Instancia de una conexión compartida</param>
        /// <returns>Último valor asignado.</returns>
        public static int RetrieveLastKeyCustom(IDbConnection connection = null)
        {
            return (int)Database.Select("SELECT NVL(MAX(UserId),0) " +
                                     "FROM UserMember WHERE UserId < 999999000")
                                .QueryScalar<Decimal>(connection, "Research");
        }
    }
}